master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,643 @@
+            backoffs = [0, 1, 5, 10, 15, 15, 15, 15, 15, 15, 15, 60, 60, 60, 60, 60, 60, 60, 60, 60]
+
+            self.ship(data, backoffs)
+
+    def ship(self, data: bytes, backoffs: List[int]) -> None:
+        for delay in backoffs:
+            time.sleep(delay)
+            try:
+                req = urllib.request.Request(self.url, data, self.headers, method="POST")
+                with urllib.request.urlopen(req, context=self.context) as resp:
+                    respbody = resp.read()
+
+                # XXX: what am I supposed to be checking here?
+
+                assert resp.getcode() == 200, (resp.get_code(), respbody.decode("utf8"))
+                # Shipped successfully
+                return
+
+            except Exception:
+                logging.error("failed to ship logs to master", exc_info=True)
+                pass
+
+        raise RuntimeError("failed to connect to master for too long, giving up")
+
+    def ship_special(self, msg: str, metadata: Dict[str, str], emit_stdout_logs: bool) -> None:
+        """
+        Ship a special message, probably from failing to start the child process.
+        """
+        now = datetime.datetime.now(datetime.timezone.utc).isoformat()
+        logs = []
+        # Build a json log line out of each message line.
+        if not msg.endswith("\n"):
+            msg += "\n"
+        for line in msg.splitlines(keepends=True):
+            if emit_stdout_logs:
+                print(line, end="", file=sys.stderr)
+            logs.append(
+                {
+                    "timestamp": now,
+                    "log": line,
+                    "level": "ERROR",
+                    "stdtype": "stderr",
+                    **metadata,
+                }
+            )
+
+        data = json.dumps(logs).encode("utf8")
+
+        # Try to ship for about 30 seconds.
+        backoffs = [0, 1, 5, 10, 15]
+        self.ship(data, backoffs)
+
+
+class Waiter(threading.Thread):
+    """
+    Waiter calls p.wait() on a process, that's it.
+    """
+
+    def __init__(self, p: subprocess.Popen, doneq: queue.Queue):
+        self.p = p
+        self.doneq = doneq
+        super().__init__()
+
+    def run(self) -> None:
+        try:
+            exit_code = self.p.wait()
+            self.doneq.put(DoneMsg("waiter", exit_code=exit_code, error=None))
+        except Exception as e:
+            self.doneq.put(DoneMsg("waiter", exit_code=None, error=e))
+
+
+def main(
+    master_url: str,
+    cert_name: str,
+    cert_file: str,
+    metadata: Dict[str, str],
+    token: str,
+    emit_stdout_logs: bool,
+    cmd: List[str],
+    log_wait_time: int,
+) -> int:
+    logq: queue.Queue = queue.Queue()
+    doneq: queue.Queue = queue.Queue()
+
+    waiter_started = False
+    stdout_started = False
+    stderr_started = False
+    shipper_started = False
+
+    # Normally we like structured concurrency; i.e. a function that owns a thread must not exit
```

Comment:
I don't think I've commented on this, yet, but this was a really helpful comment to read. Thanks for the effort, here.

Reply:
Thank you.

In this case, this is a canned explanation from the last time I tried to do the same (with our ProfilingAgent I think).  In that case, I actually did all the work described here before I realized it couldn't be done... sigh.